[Refactoring] Don't crash when converting a function to async that contains a call to init

ahoppen opened the pull request:
We were trying to retrieve the name of all function calls in the body using `getBaseIdentifier`. But calls to `init` don’t have a base identifier, just a `DeclBaseName` (which is special). Work with the `DeclBaseName` internally to prevent the crash.

Fixes rdar://78024731 [SR-14637]

ahoppen commented:
@swift-ci Please test

bnbarham reviewed (approved):
Formatting in the test is a bit off (the class init/subscript use 4 space and then the example func 2 space), but LGTM otherwise.

ahoppen commented:
That’s what you get from copy-pasting code with tabs and setting your tab size to 2. Updated the formatting in the test case.

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test macOS

ahoppen commented:
@swift-ci Please test Linux

ahoppen commented:
@swift-ci Please test Linux

ahoppen commented:
@swift-ci Please test macOS

ahoppen committed:
[Refactoring] Don't crash when converting a function to async that contains a call to init

We were trying to retrieve the name of all function calls in the body using `getBaseIdentifier`. But calls to `init` don’t have a base identifier, just a `DeclBaseName` (which is special). Work with the `DeclBaseName` internally to prevent the crash.

Fixes rdar://78024731 [SR-14637]

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test Linux

ahoppen commented:
@swift-ci Please macOS

ahoppen commented:
@swift-ci Please test macOS
